contrib/mesos/pkg/minion/tasks/task.go

```diff
@@ -0,0 +1,369 @@
+		panic("Task.run() may only be invoked once")
+	}
+	t.startProcess = initialState
+
+	defer close(t.done)
+	defer close(t.completedCh)
+
+	state := initialState
+	for state != nil {
+		next := state(t)
+		state = next
+	}
+}
+
+func (t *Task) tryComplete(tc *TaskCompletion) {
+	select {
+	case <-t.shouldQuit:
+		// best effort
+		select {
+		case t.completedCh <- tc:
+		default:
+		}
+	case t.completedCh <- tc:
+	}
+}
+
+// tryError is a convenience func that invokes tryComplete with a completion error
+func (t *Task) tryError(err error) {
+	t.tryComplete(&TaskCompletion{err: err})
+}
+
+type taskStateFn func(*Task) taskStateFn
+
+func taskShouldRestart(t *Task) taskStateFn {
+	// make our best effort to stop here if signalled (shouldQuit). not doing so here
+	// could add cost later (a process might be launched).
+
+	// sleep for a bit; then return t.startProcess
+	tm := time.NewTimer(t.RestartDelay)
+	defer tm.Stop()
+	select {
+	case <-tm.C:
+		select {
+		case <-t.shouldQuit:
+		default:
+			if t.Finished(true) {
+				select {
+				case <-t.shouldQuit:
+					// the world has changed, die
+					return nil
+				default:
+				}
+				return t.startProcess
+			}
+			// finish call decided not to respawn, so die
+			return nil
+		}
+	case <-t.shouldQuit:
+	}
+
+	// we're quitting, tell the Finished callback and then die
+	t.Finished(false)
+	return nil
+}
+
+func (t *Task) initLogging(r io.Reader) {
+	writer := t.createLogger()
+	go func() {
+		defer writer.Close()
+		_, err := io.Copy(writer, r)
+		if err != nil && err != io.EOF {
+			// using tryComplete is racy because the state machine closes completedCh and
+			// so we don't want to attempt to write to a closed/closing chan. so
+			// just log this for now.
+			log.Errorf("logger for task %q crashed: %v", t.bin, err)
+		}
+	}()
+}
+
+// notStartedTask spawns the given task and transitions to a startedTask state
+func notStartedTask(t *Task) taskStateFn {
+	log.Infof("starting task process %q with args '%+v'", t.bin, t.args)
+
+	// create command
+	cmd := exec.Command(t.bin, t.args...)
+	if _, err := cmd.StdoutPipe(); err != nil {
+		t.tryError(fmt.Errorf("error getting stdout of %v: %v", t.name, err))
+		return taskShouldRestart
+	}
+	stderrLogs, err := cmd.StderrPipe()
+	if err != nil {
+		t.tryError(fmt.Errorf("error getting stderr of %v: %v", t.name, err))
+		return taskShouldRestart
+	}
+
+	t.initLogging(stderrLogs)
+	if len(t.env) > 0 {
+		cmd.Env = t.env
+	}
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Setpgid:   true,
+		Pdeathsig: syscall.SIGKILL, // see cmdProcess.Kill
+	}
+
+	// last min check for shouldQuit here
+	select {
+	case <-t.shouldQuit:
+		t.tryError(fmt.Errorf("task execution canceled, aborting process launch"))
+		return taskShouldRestart
+	default:
+	}
+
+	if err := cmd.Start(); err != nil {
+		t.tryError(fmt.Errorf("failed to start task process %q: %v", t.bin, err))
+		return taskShouldRestart
+	}
+	log.Infoln("task started", t.name)
```

Comment:
I'm fine with this level of logging. I don't expect that tasks will restart very often. If it does happen, then I want it to be obvious in the logs.